File: py/src/tradekit/harness/journal.py
"""The decision journal: why a decision was made, keyed to the run.

Mirrors ``go/harness/journal.go``. A log line is prose; a journal entry is a
row you can query alongside the trade it produced. The ``decisions`` table is
in the shared schema, so a Python project and a Go project pointed at the same
database must write rows the other can read -- that, more than API symmetry,
is the requirement here.
"""

from __future__ import annotations

import datetime as dt
import json
import logging
from dataclasses import dataclass, field
from typing import Any

from tradekit.core.domain import InstrumentKey
from tradekit.store import Database
from tradekit.store._base import format_time, parse_time

__all__ = ["Decision", "Journal", "encode_detail"]

log = logging.getLogger("tradekit.harness.journal")


@dataclass(frozen=True, slots=True)
class Decision:
    """One considered action.

    Rejections matter as much as actions: "why did nothing happen today" is the
    harder question. A risk gate's reason goes straight into ``reason``, so
    "the kill switch stopped it" is queryable rather than buried in a log.
    """

    run_id: int
    at: dt.datetime
    key: InstrumentKey
    action: str
    """``"enter"``, ``"skip"``, ``"exit"``, ``"resize"``."""
    reason: str = ""
    detail: dict[str, Any] = field(default_factory=dict)
    """Free-form context, stored as JSON."""


def encode_detail(detail: dict[str, Any]) -> str:
    """Render a decision's detail as the JSON the row stores.

    Compact, keys sorted, HTML and non-ASCII left unescaped -- the same bytes
    Go's encoder writes. Sorted keys are not cosmetic: without them the same
    decision written by each language produces different bytes, and any future
    comparison or hash of journal rows diverges for no reason.
    """
    if not detail:
        return "{}"
    return json.dumps(detail, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


class Journal:
    """Records why a decision was made, keyed to the run."""

    def __init__(self, db: Database) -> None:
        """Journal into ``db``, which must have the shared schema applied."""
        self.db = db

    def record(self, d: Decision) -> None:
        """Write one decision.

        Recording must never fail the trade: an error here is logged and
        swallowed. An observability system that can stop trading is a
        liability, not an asset.
        """
        try:
            self.db.conn.execute(
                """
                INSERT INTO decisions (run_id, at, exchange, symbol, action, reason, detail_json)
                VALUES (?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    d.run_id,
                    format_time(d.at),
                    d.key.exchange,
                    d.key.symbol,
                    d.action,
                    d.reason,
                    encode_detail(d.detail),
                ),
            )
        except Exception:
            log.exception("harness: recording decision failed; trading continues")

    def since(self, run_id: int, frm: dt.datetime) -> list[Decision]:
        """A run's decisions made at or after ``frm``, oldest first."""
        rows = self.db.conn.execute(
            """
            SELECT at, exchange, symbol, action, reason, detail_json
            FROM decisions WHERE run_id = ? AND at >= ? ORDER BY at, id
            """,
            (run_id, format_time(frm)),
        )
        return [
            Decision(
                run_id=run_id,
                at=parse_time(r["at"]),
                key=InstrumentKey(r["exchange"], r["symbol"]),
                action=r["action"],
                reason=r["reason"],
                detail=json.loads(r["detail_json"]) if r["detail_json"] else {},
            )
            for r in rows
        ]

File: go/harness/journal.go
package harness

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/althk/tradekit/go/core/domain"
	"github.com/althk/tradekit/go/store"
)

// Journal records why a decision was made, keyed to the run.
//
// A log line is prose; a journal entry is a row you can query alongside the
// trade it produced. This is the thing no project had and every project's
// author has wanted at 9:20 on a Monday.
type Journal struct {
	DB *store.DB
}

// Decision is one considered action.
//
// Rejections matter as much as actions: "why did nothing happen today" is the
// harder question. risk.Gate already returns a reason with ErrBlocked; wiring
// that reason straight into Reason is the payoff for having made the gates
// return reasons rather than a bare boolean, because "the kill switch stopped
// it" becomes queryable rather than buried in a log.
type Decision struct {
	RunID  int64
	At     time.Time
	Key    domain.InstrumentKey
	Action string // "enter", "skip", "exit", "resize"
	Reason string
	// Detail is free-form context, stored as JSON.
	Detail map[string]any
}

// Record writes one decision.
//
// Recording must never fail the trade. The caller logs and swallows this
// error; an observability system that can stop trading is a liability, not an
// asset.
func (j *Journal) Record(ctx context.Context, d Decision) error {
	if j == nil || j.DB == nil {
		return fmt.Errorf("harness: Journal has no store")
	}
	detail, err := EncodeDetail(d.Detail)
	if err != nil {
		return fmt.Errorf("harness: encoding decision detail: %w", err)
	}
	_, err = j.DB.SQL().ExecContext(ctx, `
		INSERT INTO decisions (run_id, at, exchange, symbol, action, reason, detail_json)
		VALUES (?, ?, ?, ?, ?, ?, ?)`,
		d.RunID, d.At.Format(time.RFC3339Nano), d.Key.Exchange, d.Key.Symbol, d.Action, d.Reason, detail)
	if err != nil {
		return fmt.Errorf("harness: recording decision: %w", err)
	}
	return nil
}

// Since returns a run's decisions made at or after from, oldest first.
func (j *Journal) Since(ctx context.Context, runID int64, from time.Time) ([]Decision, error) {
	if j == nil || j.DB == nil {
		return nil, fmt.Errorf("harness: Journal has no store")
	}
	rows, err := j.DB.SQL().QueryContext(ctx, `
		SELECT at, exchange, symbol, action, reason, detail_json
		FROM decisions WHERE run_id = ? AND at >= ? ORDER BY at, id`,
		runID, from.Format(time.RFC3339Nano))
	if err != nil {
		return nil, fmt.Errorf("harness: reading decisions: %w", err)
	}
	defer func() { _ = rows.Close() }()

	var out []Decision
	for rows.Next() {
		var at, exchange, symbol, action, reason, detail string
		if err := rows.Scan(&at, &exchange, &symbol, &action, &reason, &detail); err != nil {
			return nil, fmt.Errorf("harness: scanning decision: %w", err)
		}
		d := Decision{RunID: runID, Key: domain.InstrumentKey{Exchange: exchange, Symbol: symbol}, Action: action, Reason: reason}
		if d.At, err = time.Parse(time.RFC3339Nano, at); err != nil {
			return nil, fmt.Errorf("harness: decision has an unreadable timestamp %q: %w", at, err)
		}
		if detail != "" && detail != "{}" {
			if err := json.Unmarshal([]byte(detail), &d.Detail); err != nil {
				return nil, fmt.Errorf("harness: decision has unreadable detail %q: %w", detail, err)
			}
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// EncodeDetail renders a decision's detail as the JSON the row stores: compact,
// keys sorted, HTML left unescaped.
//
// The form matters more than it looks. Python writes the same row with
// json.dumps(sort_keys=True, separators=(",", ":")), and without both sides
// agreeing the same decision produces different bytes in each language, so any
// future comparison or hash of journal rows diverges for no reason. Go sorts
// map keys already; the HTML escaping it does by default is what Python does
// not, hence the encoder.
func EncodeDetail(detail map[string]any) (string, error) {
	if len(detail) == 0 {
		return "{}", nil
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(detail); err != nil {
		return "", err
	}
	return string(bytes.TrimRight(buf.Bytes(), "\n")), nil
}
